api/provider/aws/task_delete.go

```diff
@@ -1,5 +1,49 @@
 package aws
 
-func (t *TaskProvider) Delete(TaskID string) error {
+import (
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ecs"
+)
+
+func (t *TaskProvider) Delete(taskID string) error {
+	environmentID, err := t.lookupTaskEnvironmentID(taskID)
+	if err != nil {
+		return err
+	}
+	fqEnvironmentID := addLayer0Prefix(t.Config.Instance(), environmentID)
+
+	taskARN, err := t.lookupTaskARN(taskID)
+	if err != nil {
+		return err
+	}
+
+	clusterName := fqEnvironmentID
+	if err := t.stopTask(clusterName, taskARN); err != nil {
+		return err
+	}
+
+	if err := deleteEntityTags(t.TagStore, "task", taskID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *TaskProvider) stopTask(clusterName, taskArn string) error {
+	input := &ecs.StopTaskInput{}
+	input.SetCluster(clusterName)
+	input.SetTask(taskArn)
+
+	if err := input.Validate(); err != nil {
+		return err
+	}
+
+	if _, err := t.AWS.ECS.StopTask(input); err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "ClusterNotFoundException" {
```

Comment:
Don't catch errors unless you plan on doing something meaningful with them. 

This should catch if the task does not exist, and `return nil`, since we want delete operations to be idempotent

Reply:
Unfortunately, a Task not found error is not among the error codes returned from `StopTask()`:

http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_StopTask.html#API_StopTask_Errors

It seemed that the most specific error I could catch given the choices was the ClusterNotFoundException. I guess given this, I should just simply check for a non-nil `err` from the call to `StopTask()` and return it, otherwise return `nil`?

Reply:
Or perhaps just changing the return to `nil` due to its idempotency?

Reply:
You need to actually force this codepath to execute and see what type of error is returned by aws-sdk-go. 

Some examples: 
* [Without specific error codes (good)](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/environment_delete.go#L60)
* [With specific error code (better) ](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/environment_delete.go#L98)

Reply:
Ok, given this and previous discussion about the lack of a Task Not Found, it appears the only option I have to force the codepath would be

```
if _, err := t.AWS.ECS.StopTask(input); err != nil {
    if err, ok := err.(awserr.Error); ok && err.Code() == "ClusterNotFoundException" {
        return nil
    }
}
```